### PR TITLE
chore(deps): bump luarocks to 3.7.0

### PR DESCRIPTION
**Pull request by clason**
Besides bugfixes, this brings `XDG_*` compliance and improved robustness, especially for network errors.

See https://github.com/luarocks/luarocks/blob/master/CHANGELOG.md

### CODE REVIEW

**Comment by justinmk** [comment]
Windows failed CI:

```
"D:\a\neovim\neovim\nvim-deps\luv.vcxproj" (default target) (22) ->
(CustomBuild target) -> 
  LINK : fatal error LNK1181: cannot open input file 'uv.lib' [D:\a\neovim\neovim\nvim-deps\luv.vcxproj]
  Error : Build error : Failed compiling module luv.dll [D:\a\neovim\neovim\nvim-deps\luv.vcxproj]

**Comment by clason** [comment]
Yeah, we haven't bumped LibUV on Windows for a while -- I assume with good reason. (I haven't touched Windows in decades, so I'm relying on help here.)

**Comment by clason** [comment]
Hmmm... unless for some reason luarocks now wants to build _its own_ version of `luv`?

**Comment by justinmk** [comment]
> Yeah, we haven't bumped LibUV on Windows for a while -- I assume with good reason

No reason. Our bundled libuv dep, for Windows, is a fork  

- https://github.com/neovim/neovim/wiki/Deps
- https://github.com/neovim/libuv/compare/v1.x...nvim

 , until the required changes are upstreamed (pending PRs in libuv repo, @erw7 is working on them).

Meanwhile, first step to get updated libuv on Windows would be to merge upstream master to the `nvim` branch in our fork: https://github.com/neovim/libuv/pull/14

**Comment by clason** [comment]
Well, LibUV recently dropped support for Windows 7, so that needs to be done with care...

**Comment by erw7** [comment]
@clason For now, the following patch will solve the problem of build failures with MSVC.

```diff
diff --git a/third-party/cmake/BuildLuarocks.cmake b/third-party/cmake/BuildLuarocks.cmake
index a961ec69c..2bb448333 100644
--- a/third-party/cmake/BuildLuarocks.cmake
+++ b/third-party/cmake/BuildLuarocks.cmake
@@ -202,6 +202,9 @@ if(USE_BUNDLED_BUSTED)
     set(LUV_ARGS "CFLAGS=-O0 -g3 -fPIC")
     if(USE_BUNDLED_LIBUV)
       list(APPEND LUV_ARGS LIBUV_DIR=${HOSTDEPS_INSTALL_DIR})
+      if(MSVC)
+        list(APPEND LUV_ARGS LIBUV_LIBDIR=${HOSTDEPS_INSTALL_DIR}/lib)
+      endif()
     endif()
     SET(LUV_PRIVATE_ARGS LUA_COMPAT53_INCDIR=${DEPS_BUILD_DIR}/src/lua-compat-5.3/c-api)
     add_custom_command(OUTPUT ${ROCKS_DIR}/luv
```

**Comment by clason** [comment]
@erw7 Thank you, that did the trick!

**Comment by erw7** [comment]
The cause of the build failure seems to be a bug in luerocks by https://github.com/luarocks/luarocks/commit/83126ba324846b754ffc5e0345341f01262b3f86. The `LIBUV_LIBDIR` setting is just a workaround, so please write some comments.

**Comment by clason** [comment]
@erw7 sorry, you're right. done! (or let me know if the comment can be improved.)

File: third-party/cmake/BuildLuarocks.cmake [hunk 1]
```diff
@@ -202,6 +202,11 @@ if(USE_BUNDLED_BUSTED)
     set(LUV_ARGS "CFLAGS=-O0 -g3 -fPIC")
     if(USE_BUNDLED_LIBUV)
       list(APPEND LUV_ARGS LIBUV_DIR=${HOSTDEPS_INSTALL_DIR})
+      # workaround for bug introduced in
+      # https://github.com/luarocks/luarocks/commit/83126ba324846b754ffc5e0345341f01262b3f86
```
[the remark this answers is not in the document]
Comment (clason) [reply third-party/cmake/BuildLuarocks.cmake:206]: Me neither, and the change was made in 3.3.0 already...

[the remark this answers is not in the document]
Comment (clason) [reply third-party/cmake/BuildLuarocks.cmake:206]: Might be https://github.com/luarocks/luarocks/issues/1214 though?

[the remark this answers is not in the document]
Comment (erw7) [reply third-party/cmake/BuildLuarocks.cmake:206]: I created the PR (https://github.com/luarocks/luarocks/pull/1355).

**Comment by erw7** [comment]
> Yeah, we haven't bumped LibUV on Windows for a while -- I assume with good reason.

I didn't do that because I didn't know what criteria were used to select notable changes to include in the commit message. The `nvim` branch is now up to date with `1.42.0`, so you can update `libuv` for Windows as well. Once we update to 1.40.0 or later, #15117 (Emoji with code points higher than `0x10000` are displayed as Replacement Character `U+FFFD`) should be resolved.

If we as a project are going to continue to support Windows 7, we need to make changes to do so. My position is that if users are going to use an EOL OS like Windows 7, the cost should be paid by each user. Therefore, I will not be doing any work for it.

**Comment by justinmk** [comment]
> If we as a project are going to continue to support Windows 7, we need to make changes to do so. My position is that if users are going to use an EOL OS like Windows 7, the cost should be paid by each user. Therefore, I will not be doing any work for it.

We will only support what libuv supports.

**Comment by clason** [comment]
Then I can try removing the special case for Windows in the libuv bump PR, would be nice to unify the deps again.

(Further discussion on that PR; feel free to merge this one since the LibUV discussion is not related.)

**Pull request merged**